Added BraveShieldsDisabledForUrls and BraveShieldsEnabledForUrls policies

spylogsster opened the pull request:
<!-- Add brave-browser issue below that this PR will resolve -->
Resolves https://github.com/brave/brave-browser/issues/25394

- Added `BraveShieldsDisabledForUrls` to disable shields urls
- Added `BraveShieldsEnabledForUrls` to enable shields urls

Urls set by policies are protected from editing via popup or other ways

![image](https://user-images.githubusercontent.com/2965009/195894730-8116b1a7-c234-4375-b585-50024366e697.png)

![image](https://user-images.githubusercontent.com/2965009/195894833-c9fb0107-af8b-4c73-a998-22e6b6490cdc.png)

<img width="398" alt="image" src="https://user-images.githubusercontent.com/2965009/199205216-81da3227-65e2-435a-81d1-18019b80f269.png">

<img width="387" alt="image" src="https://user-images.githubusercontent.com/2965009/199205283-df899986-1a52-45a9-931d-596fa39566c9.png">

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

- Enable policy values in `Computer\HKEY_CURRENT_USER\SOFTWARE\Policies\BraveSoftware\Brave` and check that settings set by policy cannot be changed by user and work as expected

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/brave-25394/64503e43af22d4cbcef3fbfe382921e61eac8533/storybook/index.html)

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/brave-25394/3573885472da69c02121de29cae671b92c1b8189/storybook/index.html)

bridiver commented on script/policy_source_helper.py:
```diff
@@ -87,6 +87,48 @@ def AddBravePolicies(template_file_contents):
                      '''Wallet feature will be disabled.'''),
         },
+        {
+            'name': 'ShieldsBlockedForUrls',
```
Please use consistent terminology for shields (enabled/disabled).

bridiver replied:
this applies everywhere

bridiver replied:
block/allow is very confusing when applied to shields and should only be used internally to check content settings. Everything else should reference enable/disable

bsclifton replied:
@bridiver this is matching some of what Chromium has for policy names:
- https://chromeenterprise.google/policies/#ClipboardAllowedForUrls
- https://chromeenterprise.google/policies/#ClipboardBlockedForUrls
- https://chromeenterprise.google/policies/#CookiesAllowedForUrls
- etc

bridiver replied:
@bsclifton that doesn't matter. Javascript blocked/allowed makes sense and is easy to understand. The same with cookies, etc... We need to use consistent terminology here and shields blocked/allowed is not consistent and it's extremely confusing.

bridiver replied:
to be clear I'm not talking about removing `ForUrls`, just using enabled/disabled instead of allow/block. Honestly without looking at the code I couldn't even tell you whether allow is enabled or disabled and that's the problem with using allow/block with shields

bsclifton replied:
I can see how block/allow would be confusing for shields - maybe we can change these to 
- `ShieldsEnabledForUrls`
- `ShieldsDisabledForUrls`

spylogsster replied:
renamed

goodov commented on script/policy_source_helper.py:
```diff
@@ -87,6 +87,48 @@ def AddBravePolicies(template_file_contents):
+              'type': 'array',
+              'items': { 'type': 'string' },
+            },
+            'supported_on': ['chrome.*:106-'],
```
`107-`

spylogsster replied:
done

goodov commented on script/policy_source_helper.py:
```diff
@@ -87,6 +87,48 @@ def AddBravePolicies(template_file_contents):
+              'type': 'array',
+              'items': { 'type': 'string' },
+            },
+            'supported_on': ['chrome.*:106-'],
```
`107-`

spylogsster replied:
done

goodov commented on script/policy_source_helper.py:
```diff
@@ -87,6 +87,48 @@ def AddBravePolicies(template_file_contents):
+                'can_be_recommended': False,
+                'can_be_mandatory': True
+            },
+            'example_value': True,
```
`example_value` is incorrect.

spylogsster replied:
done

goodov commented on components/content_settings/core/browser/brave_content_settings_policy_provider.cc:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    base::Value::ConstListView pattern_str_list =
+        pref->GetValue()->GetListDeprecated();
```
pls don't use `*Deprecated`.

spylogsster replied:
done

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/brave-25394/c8b902c70182cbb35b6e4b09683bfaeb95f8d440/storybook/index.html)

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/brave-25394/a53efedcd38033749e24ec58c18745c1800855d6/storybook/index.html)

bsclifton commented:
Functionally, this works perfectly! 😄 Some notes
- Created `HKEY_LOCAL_MACHINE\SOFTWARE\Policies\BraveSoftware\Brave\ShieldsBlockedForUrls`
    - created `REG_SZ` value `1` set to `https://www.example.com`
    - created `REG_SZ` value `2` set to `[*.]brave.com`
- Created `HKEY_LOCAL_MACHINE\SOFTWARE\Policies\BraveSoftware\Brave\ShieldsAllowedForUrls`
    - created `REG_SZ` value `1` set to `[*.]twitter.com`
    - created `REG_SZ` value `1` set to `https://www.example.com`
- Policies show under brave://policy
    ![image](https://user-images.githubusercontent.com/4733304/197294813-1303f096-bc20-4efb-bd32-e8ce59319f3b.png)
    ![image](https://user-images.githubusercontent.com/4733304/197294839-52d06061-9790-4b5a-a273-25aebc59f3f3.png)
- I forget if brave://management is supposed to show something? but it looks empty. Maybe this is regular?
    ![image](https://user-images.githubusercontent.com/4733304/197294903-966980d1-040e-48e0-9e64-448a5e027f32.png)
- Shields itself respect the setting like you would expect! 😄
    - User can't toggle for ones called out. For ones not in list, you can edit like usual
    - ![image](https://user-images.githubusercontent.com/4733304/197295323-00af6b3b-d764-4ada-b1a8-318c4b1a37ed.png)
    - In my examples:
        - you can't turn OFF shields for twitter.com (or any subdomain) 👍 
        - you can't turn ON shields for brave.com (or any subdomain) 👍 
        - `https://www.example.com` has both defined; but ON takes precedence 👍 
    - No tooltip shown for shields. This might be a good enhancement (ex: if you hover over, maybe it says "this setting is managed by your organization"

bsclifton commented:
Per comment from @bridiver - I do think it makes sense to change the names of the keys to: `ShieldsEnabledForUrls` and `ShieldsDisabledForUrls`. This makes things a lot more clear 😄👍 

Will check out the code now too (quick review of it looked great!)

spylogsster commented:
> Per comment from @bridiver - I do think it makes sense to change the names of the keys to: `ShieldsEnabledForUrls` and `ShieldsDisabledForUrls`. This makes things a lot more clear 😄👍
> 
> Will check out the code now too (quick review of it looked great!)

done

spylogsster commented:
> Functionally, this works perfectly! 😄 Some notes
> 
> * Created `HKEY_LOCAL_MACHINE\SOFTWARE\Policies\BraveSoftware\Brave\ShieldsBlockedForUrls`
>   
>   * created `REG_SZ` value `1` set to `https://www.example.com`
>   * created `REG_SZ` value `2` set to `[*.]brave.com`
> * Created `HKEY_LOCAL_MACHINE\SOFTWARE\Policies\BraveSoftware\Brave\ShieldsAllowedForUrls`
>   
>   * created `REG_SZ` value `1` set to `[*.]twitter.com`
>   * created `REG_SZ` value `1` set to `https://www.example.com`
> * Policies show under brave://policy
>   ![image](https://user-images.githubusercontent.com/4733304/197294813-1303f096-bc20-4efb-bd32-e8ce59319f3b.png)
>   ![image](https://user-images.githubusercontent.com/4733304/197294839-52d06061-9790-4b5a-a273-25aebc59f3f3.png)
> * I forget if brave://management is supposed to show something? but it looks empty. Maybe this is regular?
>   ![image](https://user-images.githubusercontent.com/4733304/197294903-966980d1-040e-48e0-9e64-448a5e027f32.png)
> * Shields itself respect the setting like you would expect! 😄
>   
>   * User can't toggle for ones called out. For ones not in list, you can edit like usual
>   * ![image](https://user-images.githubusercontent.com/4733304/197295323-00af6b3b-d764-4ada-b1a8-318c4b1a37ed.png)
>   * In my examples:
>     
>     * you can't turn OFF shields for twitter.com (or any subdomain) 👍
>     * you can't turn ON shields for brave.com (or any subdomain) 👍
>     * `https://www.example.com` has both defined; but ON takes precedence 👍
>   * No tooltip shown for shields. This might be a good enhancement (ex: if you hover over, maybe it says "this setting is managed by your organization"

Follow up to add a tooltip message https://github.com/brave/brave-browser/issues/26187

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/brave-25394/6b197fb57b7f22e361a2087e0dc42e7ee493e379/storybook/index.html)

bsclifton reviewed (approved):
This is amazing! 😄👍

bridiver commented on browser/ui/brave_shields_data_controller.h:
```diff
@@ -63,6 +63,7 @@ class BraveShieldsDataController
   AdBlockMode GetAdBlockMode();
   FingerprintMode GetFingerprintMode();
   CookieBlockMode GetCookieBlockMode();
+  bool IsShieldsManaged();
```
still needs consistent naming here `Shields` -> `BraveShields`
just like `GetBraveShieldsEnabled`, etc...
using inconsistent naming also makes finding things much more difficult when searching the code

spylogsster replied:
done, thanks

bridiver commented on browser/ui/webui/brave_shields/shields_panel_data_handler.cc:
```diff
@@ -157,6 +157,8 @@ void ShieldsPanelDataHandler::UpdateSiteBlockInfo() {
       active_shields_data_controller_->GetHttpRedirectsList();
   site_block_info_.is_shields_enabled =
       active_shields_data_controller_->GetBraveShieldsEnabled();
+  site_block_info_.is_shields_managed =
```
I pinged @nullhook about inconsistent usage here `is_shields` vs `GetBraveShields`. `Is` is preferred so I think we should have `IsBraveShieldsEnabled` and `is_brave_shields_enabled` and `is_brave_shields_managed`

spylogsster replied:
I removed all

bridiver commented on chromium_src/chrome/browser/policy/configuration_policy_handler_list_factory.cc:
```diff
@@ -32,6 +33,10 @@ const policy::PolicyToPreferenceMapEntry kBraveSimplePolicyMap[] = {
      brave_rewards::prefs::kDisabledByPolicy, base::Value::Type::BOOLEAN},
     {policy::key::kBraveWalletDisabled, brave_wallet::prefs::kDisabledByPolicy,
      base::Value::Type::BOOLEAN},
+    {policy::key::kShieldsDisabledForUrls, kManagedShieldsDisabledForUrls,
```
Shields -> BraveShields

bridiver commented on components/content_settings/core/browser/brave_content_settings_policy_provider.cc:
```diff
@@ -0,0 +1,144 @@
+
+BravePolicyProvider::BravePolicyProvider(PrefService* prefs)
+    : PolicyProvider(prefs) {
+  ReadManagedDefaultSettings();
```
why we are we calling these again, they're already called in the superclass

spylogsster replied:
this is not overriden function, but removed redundant`ReadManagedContentSettings` call

bridiver commented on components/content_settings/core/browser/brave_content_settings_policy_provider.cc:
```diff
@@ -0,0 +1,144 @@
+      continue;
+    }
+
+    const PrefService::Preference* pref =
```
all of this duplicated code from `GetContentSettingsFromPreferences` is not ok. Patching code is better than duplicating code so I think we can probably remove this class and just patch `kPrefsForManagedContentSettingsMap` and `kManagedPrefs` (`#define BRAVE_MANAGED_PREFS ...`)

spylogsster replied:
done

bridiver commented on components/brave_shields/resources/panel/components/main-panel/index.tsx:
```diff
@@ -135,6 +135,7 @@ function MainPanel () {
               isOn={siteBlockInfo?.isShieldsEnabled}
               onChange={handleToggleChange}
               accessibleLabel={getLocale('braveShieldsEnable')}
+              disabled={siteBlockInfo?.isShieldsManaged}
```
how is this actually displaying to the user? Is it obvious that it's disabled by policy? What happens if the user tries to click anyway? We need browser tests to check that the UI shows the correct value and behaves correctly when the setting is managed.

spylogsster replied:
you can see the picture in the description 
![image](https://user-images.githubusercontent.com/2965009/197778871-cc326c7c-2f97-458c-80cf-dc8447d6ba14.png)

bridiver commented:
what does `brave_shields::GetBraveShieldsEnabled`  return for a managed url? There should be a test to verify that it returns the correct (managed) result

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/brave-25394/be527386d8f53c6728410125c74c44c0b1e2a7cc/storybook/index.html)

spylogsster commented:
> what does `brave_shields::GetBraveShieldsEnabled` return for a managed url? There should be a test to verify that it returns the correct (managed) result. I believe it should return the correct value, but there should be a unit test to verify that

I added this cases to `BraveShieldsUtilTest, SetBraveShieldsEnabled_ForOrigin` but did it similar as others call  in test through `HostContentSettingsMap`. Changed it to use `brave_shields::GetBraveShieldsEnabled`

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/brave-25394/64e3cc2267c14fbb5b61379ea76643ffb41465f0/storybook/index.html)

bsclifton commented:
OK thanks @bridiver for raising some issues - I am going to update https://github.com/brave/brave-browser/issues/25394 to have proper documentation for how this feature should work and how the states should behave. If that becomes too complex, I'll create a spec for this

We'll also need to pull in https://github.com/brave/brave-browser/issues/26187 to make it obvious what happened. For this, @spylogsster has shared we'll need some UI help. I'm going to mark PR as DRAFT again until we resolve this 😄

bsclifton reviewed (changes requested):
Marking as `Request changes` until we have the UI changes in place (see last comment from me). I'll get more information in the original issue

bsclifton commented:
Quick update - the original issue (https://github.com/brave/brave-browser/issues/25394) has been updated (proper test plan / information)- and we also have a design here too from @aguscruiz 😄👍 
https://www.figma.com/file/3A6F6VrxVahiZFxLr7j7FO/Desktop-Shields?node-id=2695%3A49651

spylogsster commented:
> Quick update - the original issue ([brave/brave-browser#25394](https://github.com/brave/brave-browser/issues/25394)) has been updated (proper test plan / information)- and we also have a design here too from @aguscruiz 😄👍 https://www.figma.com/file/3A6F6VrxVahiZFxLr7j7FO/Desktop-Shields?node-id=2695%3A49651

Added UI message from figma and blocked controls if setting is managed and reopened to review, updated screenshots in the description

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/brave-25394/2560fde6c480851ce21c84ba1fa4d53deabe5a3e/storybook/index.html)

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/brave-25394/7061de3990145e73947c5c1d75244f69fdab1dc5/storybook/index.html)

bridiver commented on patches/components-content_settings-core-browser-content_settings_policy_provider.cc.patch:
```diff
@@ -0,0 +1,22 @@
+ 
+ constexpr PrefsForManagedContentSettingsMapEntry
+-    kPrefsForManagedContentSettingsMap[] = {
++    kPrefsForManagedContentSettingsMap[] = { BRAVE_MANAGED_CONTENT_SETTINGS
```
we shouldn't be changing any lines, only adding a new one. Same below

spylogsster replied:
done

bridiver commented on script/policy_source_helper.py:
```diff
@@ -87,6 +87,48 @@ def AddBravePolicies(template_file_contents):
+            },
+            'example_value': ['https://brave.com'],
+            'id': 4,
+            'caption': '''Blocks Brave Shields for urls.''',
```
enable/disable for all references here

spylogsster replied:
done, thanks

bridiver commented on script/policy_source_helper.py:
```diff
@@ -87,6 +87,48 @@ def AddBravePolicies(template_file_contents):
+            'id': 5,
+            'caption': '''Allows Brave Shields for urls.''',
+            'tags': [],
+            'desc': ('''This policy allows an admin to specify that Brave '''
```
enabled/disabled

spylogsster replied:
done

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/brave-25394/a157ffa807f046f6167f8b61eb05a09302372aff/storybook/index.html)

bridiver commented on components/brave_shields/browser/brave_shields_util_unittest.cc:
```diff
@@ -168,6 +171,90 @@ TEST_F(BraveShieldsUtilTest, SetBraveShieldsEnabled_ForOrigin) {
+
+  GURL blocked("http://blocked.com");
+  GURL allowed("http://allowed.com");
+  // Set allowed as blocked for initial value.
```
This is very confusing. I don't think using "allowed" and "blocked" for domains is a good idea. Also we should be using enabled/disabled as well here for everything except the actual content setting value itself

spylogsster replied:
renamed

bridiver replied:
sorry, I didn't mean change these to enable/disable, my point is that using any variation of block/allow/enable/disable is confusing because you are changing the content setting values for them. Something generic like host1.com, host2.com is better here I think.

spylogsster replied:
renamed, thanks

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/brave-25394/6b4b3f2e03b87b9b6be5d8a36b88d0b6b8466807/storybook/index.html)

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/brave-25394/1ac6b36388590015fc58d9273a911427425b0020/storybook/index.html)

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/brave-25394/83fb7d2c46676cc264c53303ecc0a72d55ccd31a/storybook/index.html)

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/brave-25394/b388d7e83bc96331d80414d1d7d86389538b816f/storybook/index.html)
